tachyon_core/src/storage/file.rs

```diff
@@ -18,6 +19,70 @@ const MAGIC: [u8; MAGIC_SIZE] = [b'T', b'a', b'c', b'h'];
 
 const HEADER_SIZE: usize = 71;
 
+pub struct FileLockGuard {
+    fd: RawFd,
+    released: bool,
+}
+
+impl FileLockGuard {
+    pub fn new_shared(fd: &File) -> Result<Self, io::Error> {
+        rustix::fs::flock(fd.as_fd(), rustix::fs::FlockOperation::LockShared)?;
+        Ok(Self {
+            fd: fd.as_raw_fd(),
+            released: false,
+        })
+    }
+
+    pub fn new_exclusive(fd: &File) -> Result<Self, io::Error> {
+        rustix::fs::flock(fd.as_fd(), rustix::fs::FlockOperation::LockExclusive)?;
+        Ok(Self {
+            fd: fd.as_raw_fd(),
+            released: false,
+        })
+    }
+
+    pub fn non_blocking_new_shared(fd: &File) -> Result<Self, io::Error> {
+        rustix::fs::flock(
+            fd.as_fd(),
+            rustix::fs::FlockOperation::NonBlockingLockShared,
+        )?;
+        Ok(Self {
+            fd: fd.as_raw_fd(),
+            released: false,
+        })
+    }
+
+    pub fn non_blocking_new_exclusive(fd: &File) -> Result<Self, io::Error> {
+        rustix::fs::flock(
+            fd.as_fd(),
+            rustix::fs::FlockOperation::NonBlockingLockExclusive,
+        )?;
+        Ok(Self {
+            fd: fd.as_raw_fd(),
+            released: false,
+        })
+    }
+
+    pub fn release(&mut self) -> Result<(), io::Error> {
```

Comment:
Could release be called concurrently? If so there could be 2 threads that set `released` to true

Reply:
Nah a connection cannot be shared among multiple threads

Reply:
Nice makes sense